### ISSUE TITLE
Wrong type is generated by typescript

### ISSUE THREAD

**Issue by yigalirani** [microsoft/TypeScript#60043]
### 🔎 Search Terms

generated type

### 🕗 Version & Regression Information

- This changed between versions ______ and _______
- This changed in commit or PR _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

https://www.typescriptlang.org/play/?moduleResolution=99&target=99&jsx=0&module=199#code/KYDwDg9gTgLgBAMwK4DsDGMCWEV0wWwApgAbALjhSXwCNgo4AfSpEkgSjgG8AoOPBHGIk4AXlEs2nXv35ocAZ3gBDFArHc+s-qExKFFBMpILgAGi38AvgG5LcKMBhIouVQrvWt8tSrUAmDRlZADdjJGAKAHYLbV19ChgoCNi4Wy1HZ1c4d387Kx4gA

### 💻 Code

```ts
declare function get_el(): number | null;
var el=get_el() 
export const im=function(){
  if (el==null){
    const ans= {
      exists:false
    }
    return ans
  }
  const ans2={
    value:7,    
    exists:true,

  }    
  return ans2
}()
```

### 🙁 Actual behavior

const im: {
    value: number;
    exists: boolean;
} | {
    exists: boolean;
}

### 🙂 Expected behavior

export declare const im: {
    exists: boolean;
};

### Additional information about the issue

see description in https://stackoverflow.com/questions/79017053/wrong-type-is-generated-by-typescript/79017129

**Comment by yigalirani** [comment]
oops the Actual behavior and Expected behavior are reversed

**Comment by MartinJohns** [comment]
This is working as intended and is due to "subtype reduction". If you don't want this you can add an explicit type annotation.

**Comment by fregante** [comment]
@MartinJohns @RyanCavanaugh I get it that it works as intended and that the user can "fix it" by changing the code, but the result is objectively wrong/lossy.

**Comment by RyanCavanaugh** [comment]
> the result is objectively wrong/lossy and there's room for improvement.

It's not objectively wrong, and the proposed behavior is actually not sound.

**Comment by fregante** [comment]
In plain English: _the function may return an object with a key named `value`_

In TypeScript: _no_

I'm sure I lack the understanding of why you would say that, but to me TypeScript currently does not match the reality.

**Comment by fregante** [comment]
Ok, maybe not "wrong" but certainly lossy. Having a return type `{x?: unknown}` would also be "correct" according to your definition, but we'd lose two keys instead of one.

I don’t think anyone looking at that code won't be surprised of the current inferred type.
